Enhancements

tomossomot opened the pull request:
hi @nkostoulas 

- [issue 111 - error on topupscript sigs](https://github.com/commerceblock/mainstay/issues/111) 
- [issue 122 - set current fee to unconfirmed tx fee upon restart](https://github.com/commerceblock/mainstay/issues/122)
- [issue 123 - attestation timing precision](https://github.com/commerceblock/mainstay/issues/123)
- [issue 89- logging system](https://github.com/commerceblock/mainstay/issues/89)

I've done the logging system to look and program similar to coordinator, but im not sure if this is a good way to do it. lmk

nkostoulas commented on attestation/attestclient.go:
```diff
@@ -155,16 +156,21 @@ func newMultisigAttestClient(config *confpkg.Config, isSigner bool, wif *btcutil
 	pubkeys, numOfSigs := crypto.ParseRedeemScript(multisig)
 
+	// Verify same amount of sigs in topupscript and init script
+	if _, topUpnumOfSigs := crypto.ParseRedeemScript(config.TopupScript()); topUpnumOfSigs != topUpnumOfSigs {
```
`topUpnumOfSigs != topUpnumOfSigs` 👍

nkostoulas commented on db/db_mongo.go:
```diff
@@ -422,14 +422,14 @@ func (d *DbMongo) GetAttestationMerkleCommitments(txid chainhash.Hash) ([]models
 		var commitmentDoc bsonx.Doc
 		if err := res.Decode(&commitmentDoc); err != nil {
-			fmt.Printf("%s\n", BadDataMerkleCommitmentCol)
+			log.Infof("%s\n", BadDataMerkleCommitmentCol)
```
warning

nkostoulas commented on db/db_mongo.go:
```diff
@@ -422,14 +422,14 @@ func (d *DbMongo) GetAttestationMerkleCommitments(txid chainhash.Hash) ([]models
 		modelErr := models.GetModelFromDocument(&commitmentDoc, commitmentModel)
 		if modelErr != nil {
-			fmt.Printf("%s\n", BadDataMerkleCommitmentCol)
+			log.Infof("%s\n", BadDataMerkleCommitmentCol)
```
warning

nkostoulas commented on attestation/attestservice.go:
```diff
@@ -127,16 +127,22 @@ func NewAttestService(ctx context.Context, wg *sync.WaitGroup, server *AttestSer
+	log.Infof("Time handle unconfirmed set to: %v\n", atimeHandleUnconfirmed)
+
+	// If current unconfirmed tx exists set fee to unconfirmed tx's fee (mainstay restart)
+	if unconfirmed, unconfirmedTxid, _ := attester.getUnconfirmedTx(); unconfirmed {
```
This should be moved to `stateInitUnconfirmed`.

nkostoulas replied:
We get the unconfirmed tx there already.

nkostoulas replied:
And the raw transaction too so you don't need `GetMempoolEntry` either ( I think )

tomossomot replied:
`GetTransaction` doesnt get fee or amount for unconfirmed txs for some reason. Maybe that info isnt considered set until its validated by a block producer? 
Changed `GetTransaction` to `GetMempoolEntry` as this rpc has everything the code needs. Cant get `Time` or `Fee` from `GetRawTransaction` easily

nkostoulas replied:
Hmm that's true. I checked and it seems that `GetTransaction` only has fee information for transactions that are spendable by our wallet (which is probably not the case in the unit-tests and definitely not the case in our production with HSMs) or watch only (if you add the watch only flag which seems to be disabled in the btcd) library.

nkostoulas commented on attestation/attestfees.go:
```diff
@@ -129,20 +129,26 @@ func (a *AttestFees) ResetFee(useMinimum ...bool) {
 }
 
+// Manually force set current fee regardless of max, min values
+func (a *AttestFees) setCurrentFee(fee int) {
```
Can you add a small unit-test for this in attestfees_test.go?

nkostoulas commented on attestation/attestclient.go:
```diff
@@ -103,11 +104,11 @@ func parseTopupKeys(config *confpkg.Config, isSigner bool) *btcutil.WIF {
 			return topupWif
 		} else {
-			log.Println(WarningTopupPkMissing)
+			log.Infoln(WarningTopupPkMissing)
```
warning

nkostoulas commented on attestation/attestclient.go:
```diff
@@ -229,14 +235,14 @@ func NewAttestClient(config *confpkg.Config, signerFlag ...bool) *AttestClient {
 		pkWifTopup = parseTopupKeys(config, isSigner)
 	} else {
-		log.Println(WarningTopupInfoMissing)
+		log.Infoln(WarningTopupInfoMissing)
```
warning

nkostoulas commented on attestation/attestclient.go:
```diff
@@ -397,7 +403,7 @@ func (w *AttestClient) createAttestation(paytoaddr btcutil.Address, unspent []bt
 	// print warning if txout value less than 100*maxfee target
 	if msgTx.TxOut[0].Value < 100*maxFee {
-		log.Println(WarningInsufficientFunds)
+		log.Infoln(WarningInsufficientFunds)
```
warning

nkostoulas commented on attestation/attestclient.go:
```diff
@@ -521,7 +527,7 @@ func (w *AttestClient) getTransactionPreImages(hash chainhash.Hash, msgTx *wire.
 		topupScriptSer, topupDecodeErr := hex.DecodeString(w.scriptTopup)
 		if topupDecodeErr != nil {
-			log.Printf("%s %s\n", WarningFailedDecodingTopupMultisig, w.scriptTopup)
+			log.Infof("%s %s\n", WarningFailedDecodingTopupMultisig, w.scriptTopup)
```
warning

nkostoulas commented on attestation/attestclient.go:
```diff
@@ -229,14 +235,14 @@ func NewAttestClient(config *confpkg.Config, signerFlag ...bool) *AttestClient {
 		importErr := config.MainClient().ImportAddressRescan(topupAddrStr, "", false)
 		if importErr != nil {
-			log.Printf("%s (%s)\n%v\n", WarningFailureImportingTopupAddress, topupAddrStr, importErr)
+			log.Infof("%s (%s)\n%v\n", WarningFailureImportingTopupAddress, topupAddrStr, importErr)
```
warning

nkostoulas commented on attestation/attestservice_test.go:
```diff
@@ -147,8 +147,6 @@ func verifyStateHandleUnconfirmedToSignAttestation(t *testing.T, attestService *
 	assert.Equal(t, 1, len(attestService.attestation.Tx.TxOut))
 	assert.Equal(t, 0, len(attestService.attestation.Tx.TxIn[0].SignatureScript))
 	assert.Equal(t, ATimeSigs, attestDelay)
-	assert.Equal(t, attestService.attester.Fees.minFee+attestService.attester.Fees.feeIncrement,
```
Just because your change broke all the tests you shouldn't remove this. It tests that when an attestation is unconfirmed the fee is bumped correctly in **all** unit-tests. I assume that is because the tests no longer use minFee in the first attestation (which is the default) but use an override fee with `setCurrentFee`.

tomossomot replied:
The problem here was that AStateInit -> AStateAwaitConfirmation now resets fees with changes in this PR. Doing the manual `BumpFee()` after this state change fixes this.

nkostoulas commented on attestation/attestservice.go:
```diff
@@ -196,8 +196,10 @@ func (s *AttestService) stateInitUnconfirmed(unconfirmedTxid chainhash.Hash) {
+	walletTx, _ := s.config.MainClient().GetMempoolEntry(unconfirmedTxid.String())
 	confirmTime = time.Unix(walletTx.Time, 0)
+
+	s.attester.Fees.setCurrentFee(int(walletTx.Fee*100000000)) // Set fee to unconfirmed tx's fee
```
you can replace 100000000 with attest client COIN

nkostoulas replied:
Actually the calculation is wrong. AttestFees stores fee per byte not absolute fee.

nkostoulas replied:
You need the transaction size to calculate fee per byte. When creating the transaction we can only estimate this size, since signatures can vary from 71-73 bytes and given that using multisig that has >1 sigs and that when the script size is larger than `0xfd` another 2 bytes are introduced, means that the actual size might potentially be off by 4 bytes (I think). For this you can check `attestclient:calcSignedTxSize()`. 

This is fine **but** you might not be able to unit-test with exact values but only approximate.

nkostoulas commented on attestation/attestservice_test.go:
```diff
@@ -827,12 +830,18 @@ func TestAttestService_FailureSendAttestation(t *testing.T) {
+		_, unconfirmedTxid, _ := attestService.attester.getUnconfirmedTx()
+		tx, _ := config.MainClient().GetMempoolEntry(unconfirmedTxid.String())
+		assert.Equal(t, attestService.attester.Fees.GetFee(), int(tx.Fee*100000000))
+
```
When you are testing you want to check the value you set was correct not replicate what you do in the code.
This:
`assert.Equal(t, attestService.attester.Fees.GetFee(), int(tx.Fee*100000000))`
is the same as your code.

This:
`assert.Equal(t, attestService.attester.Fees.GetFee(), 23)`
would be a correct test.

Which would fail because you are setting it to the absolute fee not fee per byte (as per previous comment).

tomossomot commented:
- included fee inheritance from unconfirmed in attesteService.Fees
- Added function to compare `models.AttestationInfo` whose Amount fields may be slightly off because of imprecise feePerByte calculations
- seperate function for testing of unconfirmed tx fee grabbing for simplicity
- removed 

`assert.Equal(t, attestService.attester.Fees.minFee+attestService.attester.Fees.feeIncrement,
			attestService.attester.Fees.GetFee())`

since the same test in performed in `verifyStateHandleUnconfirmedToSignAttestation(t, attestService)`

nkostoulas commented on attestation/attestservice_test.go:
```diff
@@ -830,18 +843,12 @@ func TestAttestService_FailureSendAttestation(t *testing.T) {
-		// Test new fee set to unconfirmed tx's fee after restart
-		_, unconfirmedTxid, _ := attestService.attester.getUnconfirmedTx()
-		tx, _ := config.MainClient().GetMempoolEntry(unconfirmedTxid.String())
-		assert.Equal(t, attestService.attester.Fees.GetFee(), int(tx.Fee*100000000))
```
What I meant in my last review is that you might set the fee to 23 but the actual fee set might be 29.8, so when you did this check it would fail. Transaction amounts don't change though.

nkostoulas replied:
You don't need a new test case for this. Just do the check from `TestAttestService_FailurePickUpUnconfirmedTxFee`, here.

tomossomot commented:
Thanks, updates made

nkostoulas commented on attestation/attestservice.go:
```diff
@@ -196,8 +196,12 @@ func (s *AttestService) stateInitUnconfirmed(unconfirmedTxid chainhash.Hash) {
 
 	s.state = AStateAwaitConfirmation // update attestation state
-	walletTx, _ := s.config.MainClient().GetTransaction(&unconfirmedTxid)
+	walletTx, _ := s.config.MainClient().GetMempoolEntry(unconfirmedTxid.String())
```
I thought I mentioned this but probably forgot. You need handle the error here because the tx might no longer be in the mempool and then you will be accessing a null pointer.

nkostoulas commented on attestation/attestservice_test.go:
```diff
@@ -937,11 +952,14 @@ func TestAttestService_FailureAwaitConfirmation(t *testing.T) {
-		Blockhash: walletTx.BlockHash,
-		Amount:    rawTx.MsgTx().TxOut[0].Value,
-		Time:      walletTx.Time}, attestService.attestation.Info)
+	assert.Equal(t,
```
use `go fmt`

nkostoulas commented on attestation/attestservice_test.go:
```diff
@@ -1079,12 +1095,64 @@ func TestAttestService_FailureHandleUnconfirmed(t *testing.T) {
+	verifyStateInitToAwaitConfirmation(t, attestService, config, latestCommitment, txid)
+
+	// Test new fee set to unconfirmed tx's feePerByte value (~23) after restart
+	assert.GreaterOrEqual(t, attestService.attester.Fees.GetFee(), 18)	// In AttestFees
```
It's max 4 bytes difference. 23±0.2 should be enough.

nkostoulas commented on attestation/attestservice_test.go:
```diff
@@ -1079,12 +1095,64 @@ func TestAttestService_FailureHandleUnconfirmed(t *testing.T) {
+
+// Test Attest Service failre
+// Test unconfirmed tx fee picked up after failure
+func TestAttestService_FailurePickUpUnconfirmedTxFee(t *testing.T) {
```
You don't need a new test case for this. You can incorporate this in `TestAttestService_FailureSendAttestation` as the only thing you are testing is checking the fee when mainstay fails after sending a new attestation. Just manually set the fee in 1 of the 3 attestations.

tomossomot replied:
Ah I misunderstood your previous comment about this.
